### ISSUE TITLE
Zotero-OPDS recognizes PDF only, not ePub attachments

### ISSUE THREAD

**Issue by mhl66** [AllThatIsTheCase/zotero-opds#1]
When accessing the embedded web server with the OPDS feed, only PDF files are visible as attachments, and HTML, tested with Lucifox (=Firefox Plugin) and PocketBook reader (=Android app)...

**Comment by retorquere** [comment]
Using Lucifox, I can see and open ePub files without issue. HTML files are indeed excluded, as most of the times I encountered them they were site snapshots, and those are not single-file documents; I don't think OPDS supports multi-file (css, js) attachments.

**Comment by mhl66** [comment]
Thank you for your quick reply, your Zotero work is much appreciated.

But sorry, only PDF attachments are shown in Lucifox (for Windows) and PocketBook: I made a collection with two entries, one PDF and one ePub, and only the first is shown in Lucifox, see attached screenshots. For PocketBook, it´s the samee.

Moreover, when I open the PDF in Firefox (via Lucifox), a file "item" is downloaded, without extension, but not opened.

Your consideration in this matter is much appreciated.
![screenshot-zotero](https://cloud.githubusercontent.com/assets/13404029/18471289/6ac12e24-79b1-11e6-94d5-82b4556a9315.png)
![screenshot-lucifox](https://cloud.githubusercontent.com/assets/13404029/18471295/73458464-79b1-11e6-8ee6-b9d9f3c67652.png)

**Comment by retorquere** [comment]
Why'd you close the issue? Is the problem resolved in some way?

**Comment by mhl66** [comment]
Sorry, sorry, this was by mistake, no, not resolved, unfortunately...

**Comment by retorquere** [comment]
Can you do the following:
1. Install https://github.com/retorquere/zotero-opds/releases/download/builds/zotero-opds-server-0.0.13-circle-5.xpi
2. Enable `Preferences`->`Advanced`->`Debug Output Logging`->`Enable`
3. Browse to that collection in Lucifox or PocketBook reader
4. Type `zotero://debug` in your address bar and press enter.
5. Copy the resulting page into a file named `opds-1-1.log` and upload it [here](http://retorquere.github.io/error-report.html) (no feedback will be given in case of success, sorry)

**Comment by mhl66** [comment]
...done... Good luck! And thank you!

**Comment by retorquere** [comment]
The epub is in the feed that is sent out, looks like it doesn't like the `application/octet-stream` mimetype. What version of Lucifox are you on? 0.9.13 picks it up just fine.

Can you try the above procedure again with [this version](https://github.com/retorquere/zotero-opds/releases/download/builds/zotero-opds-server-0.0.13-circle-6.xpi), with logfile `opds-1-2.log`?

**Comment by mhl66** [comment]
Regarding second error: Lucifox (Windows), indeed, is on 0.9.13, the mimetype (file type: pdf) seems not to be recognised, still does not open... PocketBook (Android), on the other hand, does this without any problem.

Regarding first error: still only showing entries with PDF attachments (see screenshots posted above).

Upload done, good luck again...

**Comment by retorquere** [comment]
It isn't a fallback of some kind, Zotero just claims it knows it's `application/octet-stream`. Can you try [this version](https://github.com/retorquere/zotero-opds/releases/download/builds/zotero-opds-server-0.0.13-circle-7.xpi), with (unsurprisingly I guess) `opds-1-3.log`?

**Comment by retorquere** [comment]
The reason that my lucifox picks up the epub could be because on my system, epubs are reported by Zotero to have the correct mimetype. No idea why this should be different on your system.  Seems weird this would be platform related (you use Windows, I'm currently on OSX).

**Comment by mhl66** [comment]
...done... now Lucifox says "Search failed"!?

**Comment by retorquere** [comment]
Man, why are there such important differences between Firefox on OS and Windows?! Please try https://github.com/retorquere/zotero-opds/releases/download/builds/zotero-opds-server-0.0.13-circle-8.xpi -- I've just hard-coded the mimetype for epubs.

**Comment by mhl66** [comment]
Greeeaaat!!! ePubs now showing!
![screenshot-epubworking](https://cloud.githubusercontent.com/assets/13404029/18507172/7b5d4ab0-7a6f-11e6-930a-02ec9d5fabb5.png)
Unfortunately, there is still the problem cited in the third posting here: Moreover, when I open the PDF in Firefox (via Lucifox), a file "item" is downloaded, without extension, but not opened... sorry...

**Comment by mhl66** [comment]
By the way, I´m using the standalone version of Zotero, not the Firefox one.

**Comment by retorquere** [comment]
You can try [this version](https://github.com/retorquere/zotero-opds/releases/download/builds/zotero-opds-server-0.0.13-circle-9.xpi), which puts the filename at the end of the aquisition url, but only as a query parameter. I can't make an actual url with a filename extension within Zotero.

The OPDS spec isn't explicit on the matter, but as far as I can tell the URL is inconsequential; the mimetype passed determines what the client should do. I'd recommend contacting the Lucifox/Pocketbook authors to see what they expect, and I could see why zotero standalone would force a download rather than a display (it could be missing most of the rendering stuff outside the Zotero code), but I really don't know. It's certainly not Zotero-OPDS triggering the download, I only offer an download URL, what the OPDS client does with that URL (download & render, download & save, something else entirely) is fully outside my control.

**Comment by mhl66** [comment]
Short: it works. Incredible. You are great.
Long: After some rethinking, I opened the PDF link in Internet Explorer, which displayed the PDF file just fine. Then I changed my Firefox application setting for PDF files, to open it with Acrobat, and bingo... sorry for the confusion. - Thank you! Thank you! And (did I already mention it?) thank you!

**Comment by retorquere** [comment]
My pleasure. I've cut a new release (0.0.14) which is the same as your last version so that others will also get these fixes. You'll be asked to upgrade at some point, but there won't be anything new.

**Comment by mhl66** [comment]
Great!!! What is the "Update Address" to be entered in the Zotero-OPDS setting window?

**Comment by retorquere** [comment]
That's for the [DYnDNS](https://github.com/retorquere/zotero-opds#dyndns-support) support. Please do note the potential security implications -- it seems harmless to me given how it's being used, but I'm not a security researcher.

It's only really useful if you want to sync with a device that's not the same system as the one Zotero is running on. On the same system, `localhost` will always work. It's if you have a different device, like a tablet, that it might be useful.

**Comment by mhl66** [comment]
Sorry, not the field "hostname", but the field "Update Address" I was interested in...???

**Comment by retorquere** [comment]
That is the DynDNS (or similar service) update URL. The service I used back then said that the update url should look like

```
https://www.duckdns.org/update?domains=<hostname>&token=a7c4d0ad-114e-40ef-ba1d-d217904a50f2&ip=<ip>
```

or somesuch; the OPDS plugin will replace `<hostname>` with whatever is in the `hostname` field, and `<ip>` by the detected IP address of the Zotero system.

The "new secret" button there is useless BTW; at some point it generated a TOTP secret, which ameliorated some of the potential rebinding problems, but none of the OPDS clients I found would support it. That button is now removed.
